### DISCUSSION TITLE
How do I use .java files and recompile the code?

### DISCUSSION THREAD

**Discussion by jungltemple** [MaxPixelStudios/MinecraftDecompiler#161] [Q&A] [answered]
Yeah, I don't really understand Java that much.

How do I decompile the code into .java files rather than .class files, so that I can edit them directly in IntelliJ?
And how do I run the decompiled code? It doesn't come with Gradle by default, so I'm a bit confused.
Also, for some reason my decompiled 1.21.1 code doesn't have the `net.minecraft.client.Main` class, which is the entrypoint listed in the `MANIFEST.MF` file.

**Comment by XiaoPangxie732** [comment, marked as answer]
> How do I decompile the code into .java files rather than .class files, so that I can edit them directly in IntelliJ?

Run MCD with your args and `-r -d`

> And how do I run the decompiled code? It doesn't come with Gradle by default, so I'm a bit confused.

You need to figure out it yourself. Providing a useable workspace is not the goal of MCD. And usually the decompiled code cannot be directly recompiled because it contains errors. Maybe you should use [MCP-Reborn](https://github.com/Hexeption/MCP-Reborn) instead.



> Also, for some reason my decompiled 1.21.1 code doesn't have the `net.minecraft.client.Main` class, which is the entrypoint listed in the `MANIFEST.MF` file.

It is probably a bug, I'll figure it out asap

**Comment by XiaoPangxie732** [comment]
> It is probably a bug, I'll figure it out asap

I get it. It is a bug from Mojang. The actual main class is `net.minecraft.client.main.Main`

**Comment by jungltemple** [comment]
Thank you! My problem with MCP-Reborn is that it doesn't support 1.21.1 and also has trouble with my Java 21 toolchain, but I'll think about it.
